core-ui/src/shared/ResourceForm/ResourceForm.js

```diff
@@ -0,0 +1,126 @@
+import * as jp from 'jsonpath';
+import { ModeSelector } from './components/ModeSelector';
+import { Editor } from './components/Editor';
+import { Presets } from './components/Presets';
+import * as FormComponents from './components/FormComponents';
+import './ResourceForm.scss';
+
+ResourceForm.Label = FormComponents.Label;
+ResourceForm.Input = FormComponents.Input;
+ResourceForm.CollapsibleSection = FormComponents.CollapsibleSection;
+ResourceForm.FormField = FormComponents.FormField;
+ResourceForm.TextArrayInput = FormComponents.TextArrayInput;
+ResourceForm.K8sNameField = FormComponents.K8sNameField;
+ResourceForm.KeyValueField = FormComponents.KeyValueField;
+
+export function ResourceForm({
+  pluralKind, // used for the request path
+  nameSingular,
+  resource,
+  setResource,
+  onCreate,
+  onChange,
+  formElementRef,
+  children,
+  renderEditor,
+  createFn,
+  presets,
+  onPresetSelected,
+}) {
+  const notification = useNotification();
+  const { t } = useTranslation();
+  const [mode, setMode] = React.useState(ModeSelector.MODE_SIMPLE);
+
+  if (!onCreate) {
+    onCreate = async () => {
+      try {
+        await createFn();
+        notification.notifySuccess({
+          content: t('common.create-form.messages.success', {
+            resourceType: nameSingular,
+          }),
+        });
+        LuigiClient.linkManager()
+          .fromContext('namespace')
+          .navigate(
+            `/${pluralKind.toLowerCase()}/details/${resource.metadata.name}`,
+          );
+      } catch (e) {
+        console.error(e);
+        notification.notifyError({
+          content: t('common.create-form.messages.failure', {
+            resourceType: nameSingular,
+            error: e.message,
+          }),
+        });
+        return false;
+      }
+    };
+  }
+
+  const presetsSelector = presets?.length && (
+    <Presets
+      presets={presets}
+      onSelect={({ value }) => {
+        if (onPresetSelected) {
+          onPresetSelected(value);
+        } else {
+          setResource(value);
+        }
+        onChange(new Event('input', { bubbles: true }));
+      }}
+    />
+  );
+
+  const renderFormChildren = (children, isAdvanced) =>
+    React.Children.map(children, child => {
+      if (child.props.simple && isAdvanced) {
+        return null;
+      }
+      if (child.props.advanced && !isAdvanced) {
+        return null;
+      }
+      if (!child.props.propertyPath) {
+        return child;
+      }
+      return React.cloneElement(child, {
+        value: jp.value(resource, child.props.propertyPath),
+        setValue: value => {
+          jp.value(resource, child.props.propertyPath, value);
+          setResource({ ...resource });
+        },
+      });
+    });
+
+  let editor = <Editor resource={resource} setResource={setResource} />;
+  editor = renderEditor
+    ? renderEditor({ defaultEditor: editor, Editor })
+    : editor;
+
+  return (
+    <article className="resource-form">
```

Comment:
a bit of a nitpick, but i'm not sure if `article` is an appropriate tag here, as it's semantically meant to be used for actual text articles. I think `section` would suffice.

Reply:
Well, https://developer.mozilla.org/en-US/docs/Web/HTML/Element/article gives `interactive widget` as an example, so _not only_ text articles, but well, this form is still too far from a "widget". `section` it is!